pkg/pillar/hypervisor/kvm.go

```diff
@@ -929,12 +939,184 @@ func getFmlCustomResolution(status *types.DomainStatus, globalConfig *types.Conf
+	devs      []*pciDevice
+}
+
+type multifunctionDevs map[string]*pciDevicesWithBridge // key: pci long without function number
+
+func (md multifunctionDevs) isMultiFunction(p pciDevice) bool {
+	pciLongWOFunc, err := p.pciLongWOFunction()
+	if err != nil {
+		return false
+	}
+	devs, found := md[pciLongWOFunc]
+	return found && len(devs.devs) > 1
+}
+
+func (md multifunctionDevs) index(p pciDevice) int {
+	pciLongWOFunc, err := p.pciLongWOFunction()
+	if err != nil {
+		logrus.Warnf("retrieving pci address without function failed: %v", err)
+		return -1
+	}
+	pciDevices, found := md[pciLongWOFunc]
+	if !found {
+		return -1
+	}
+
+	for i, dev := range pciDevices.devs {
+		if p.ioType == dev.ioType && p.pciLong == dev.pciLong {
+			return i
+		}
+	}
+
+	return -1
+}
+
+func multifunctionDevGroup(pcis []pciDevice) multifunctionDevs {
+	mds := make(multifunctionDevs)
+
+	for i, pa := range pcis {
+		pciWithoutFunction, err := pa.pciLongWOFunction()
+		if err != nil {
+			logrus.Warnf("retrieving pci address without function failed: %v", err)
+			continue
+		}
+
+		_, found := mds[pciWithoutFunction]
+		if !found {
+			mds[pciWithoutFunction] = &pciDevicesWithBridge{
+				bridgeBus: "",
+				devs:      []*pciDevice{},
+			}
+		}
+		mds[pciWithoutFunction].devs = append(mds[pciWithoutFunction].devs, &pcis[i])
+	}
+
+	return mds
+}
+
+func (p pciDevice) pciLongWOFunction() (string, error) {
+	pciLongSplit := strings.Split(p.pciLong, ".")
+	if len(pciLongSplit) == 0 {
+		return "", fmt.Errorf("could not split %s", p.pciLong)
+	}
+	pciWithoutFunction := strings.Join(pciLongSplit[0:len(pciLongSplit)-1], ".")
+
+	return pciWithoutFunction, nil
+}
+
+type pciAssignmentsTemplateFiller struct {
+	multifunctionDevices multifunctionDevs
+	file                 io.Writer
+}
+
+func (f *pciAssignmentsTemplateFiller) pciEBridge(pciID int, pciWOFunction string) error {
+	pciChildTemplateVars := struct {
+		PCIId int
+		Bus   int
+	}{
+		PCIId: 0,
+		Bus:   pciID,
+	}
+
+	tPCIeBridge, err := template.New("qemuPCIeBridge").Parse(qemuPCIPassthruBridgeTemplate)
+	if err != nil {
+		return fmt.Errorf("parsing qemuPCIPassthruBridgeTemplate failed: %w", err)
+	}
+	if err := tPCIeBridge.Execute(f.file, pciChildTemplateVars); err != nil {
+		return fmt.Errorf("can't write PCIe bridge Passthrough to config file (%w)", err)
+	}
+	f.multifunctionDevices[pciWOFunction].bridgeBus = fmt.Sprintf("pcie-bridge.%d", pciID)
+
+	return nil
+}
+
+func (f *pciAssignmentsTemplateFiller) do(file io.Writer, pciAssignments []pciDevice, pciID int) error {
+	if len(pciAssignments) == 0 {
+		return nil
+	}
+
+	pciPTContext := struct {
+		PCIId        int
+		PciShortAddr string
+		Xvga         bool
+		Xopregion    bool
+		Bus          string
+		Addr         string
+	}{PCIId: pciID, PciShortAddr: "", Xvga: false, Xopregion: false}
+
+	tPCI, _ := template.New("qemuPCI").Parse(qemuPciPassthruTemplate)
+
+	pciEBridgeForMultiFuncDevCreated := make(map[string]struct{}) // key: pciDevice.pciLong
```

Comment:
Is the comment correct? I see later in the code that the key is PCI Long without function.

Reply:
Fixed.
First I used `pciLong` but then the test failed and I fixed it, but I forgot to change the comment.